File: Tallest_Billboard.cpp
class Solution {
public:
    int tallestBillboard(vector<int>& rods) {
        std::unordered_map<int, int> dp;
        dp[0] = 0;

        for (int rod : rods) {
            std::unordered_map<int, int> currDP(dp);

            for (const auto& entry : dp) {
                int diff = entry.first;
                int highest = entry.second;

                int leftDiff = diff + rod;
                currDP[leftDiff] = std::max(currDP[leftDiff], highest + rod);

                int rightRod = highest - diff + rod;
                int rightDiff = std::abs(highest - rightRod);
                currDP[rightDiff] = std::max(currDP[rightDiff], std::max(highest, rightRod));
            }

            dp = currDP;
        }

        return dp[0];        
    }
};

File: Tallest_Billboard.py
class Solution:
    def tallestBillboard(self, rods: List[int]) -> int:
        dp = {0:0}

        for rod in rods:
            for diff, highest in list(dp.items()):
                left_diff = diff + rod
                dp[left_diff] = max(dp.get(left_diff, 0), highest + rod)

                right_rod = highest - diff + rod
                right_diff = abs(highest - right_rod)
                dp[right_diff] = max(dp.get(right_diff, 0), max(highest, right_rod))
        
        return dp[0]
